src/phpDocumentor/Dsn.php

```diff
@@ -0,0 +1,232 @@
+
+    /** @var string */
+    private $path;
+
+    /** @var string[]  */
+    private $query = [];
+
+    /** @var string[]  */
+    private $parameters = [];
+
+    /**
+     * Initializes the Dsn
+     *
+     * @param string $dsn
+     */
+    public function __construct($dsn)
+    {
+        $this->parse($dsn);
+    }
+
+    /**
+     * Returns the scheme part of the DSN
+     *
+     * @return string
+     */
+    public function getScheme()
+    {
+        return $this->scheme;
+    }
+
+    /**
+     * Returns the host part of the DSN
+     *
+     * @return string
+     */
+    public function getHost()
+    {
+        return $this->host;
+    }
+
+    /**
+     * Returns the port part of the DSN
+     *
+     * @return int
+     */
+    public function getPort()
+    {
+        return $this->port;
+    }
+
+    /**
+     * Returns the username part of the DSN
+     *
+     * @return string
+     */
+    public function getUsername()
+    {
+        return $this->user;
+    }
+
+    /**
+     * Returns the password part of the DSN
+     *
+     * @return string
+     */
+    public function getPassword()
+    {
+        return $this->password;
+    }
+
+    /**
+     * Returns the path part of the DSN
+     *
+     * @return string
+     */
+    public function getPath()
+    {
+        return new Path($this->path);
+    }
+
+    /**
+     * Returns the query part of the DSN
+     *
+     * @return string[]
+     */
+    public function getQuery()
+    {
+        return $this->query;
+    }
+
+    /**
+     * Returns the parameters part of the DSN
+     *
+     * @return string[]
+     */
+    public function getParameters()
+    {
+        return $this->parameters;
+    }
+
+    /**
+     * Parses the given DSN
+     *
+     * @param string $dsn
+     */
+    private function parse($dsn)
+    {
+        if (! is_string($dsn)) {
+            throw new \InvalidArgumentException(
+                sprintf('"%s" is not a valid DSN.', $dsn)
+            );
+        }
+
+        $dsnParts = explode(';', $dsn);
+        $location = $dsnParts[0];
+        unset($dsnParts[0]);
+        $locationParts = parse_url($location);
+
+        if (! isset($locationParts['scheme'])) {
+            $locationParts['scheme'] = 'file';
+            $location = 'file://' . $location;
+        }
+
+        if (! filter_var($location, FILTER_VALIDATE_URL)) {
+            throw new \InvalidArgumentException(
+                sprintf('"%s" is not a valid DSN.', $dsn)
+            );
+        }
+
+        $this->parseScheme($locationParts);
+
+        $this->parseHostAndPath($locationParts);
+
+        $this->port = isset($locationParts['port']) ? $locationParts['port'] : "";
+
+        $this->user = isset($locationParts['user']) ? $locationParts['user'] : "";
+
+        $this->password = isset($locationParts['pass']) ? $locationParts['pass'] : "";
+
+        $this->query = isset($locationParts['query']) ? explode('?', $locationParts['query']) : "";
```

Comment:
Can you change this so that `$this->query` contains an associative array where the key is the query parameter key and the name is the value?

For example:
  
    http://github.com/phpDocumentor/phpDocumentor2?milestone=3.0,query=abc

Should result in a query array that looks like this:

```
$query = [
    'milestone' => '3.0',
    'query' => 'abc'
];
```